src/snowflake/snowpark/_internal/server_connection.py

```diff
@@ -695,24 +695,31 @@ def _fix_pandas_df_integer(
         if (
             FIELD_ID_TO_NAME.get(column_metadata.type_code) == "FIXED"
             and column_metadata.precision is not None
-            and column_metadata.scale == 0
-            and not str(pandas_dtype).startswith("int")
         ):
-            # When scale = 0 and precision values are between 10-20, the integers fit into int64.
-            # If we rely only on pandas.to_numeric, it loses precision value on large integers, therefore
-            # we try to strictly use astype("int64") in this scenario. If the values are too large to
-            # fit in int64, an OverflowError is thrown and we rely on to_numeric to choose and appropriate
-            # floating datatype to represent the number.
-            if column_metadata.precision > 10:
-                try:
-                    pd_df[pandas_col_name] = pd_df[pandas_col_name].astype("int64")
-                except OverflowError:
+            if column_metadata.scale == 0 and not str(pandas_dtype).startswith("int"):
+                # When scale = 0 and precision values are between 10-20, the integers fit into int64.
+                # If we rely only on pandas.to_numeric, it loses precision value on large integers, therefore
+                # we try to strictly use astype("int64") in this scenario. If the values are too large to
+                # fit in int64, an OverflowError is thrown and we rely on to_numeric to choose and appropriate
+                # floating datatype to represent the number.
+                if column_metadata.precision > 10:
+                    try:
+                        pd_df[pandas_col_name] = pd_df[pandas_col_name].astype("int64")
+                    except OverflowError:
+                        pd_df[pandas_col_name] = pandas.to_numeric(
+                            pd_df[pandas_col_name], downcast="integer"
+                        )
+                else:
                     pd_df[pandas_col_name] = pandas.to_numeric(
                         pd_df[pandas_col_name], downcast="integer"
                     )
-            else:
-                pd_df[pandas_col_name] = pandas.to_numeric(
-                    pd_df[pandas_col_name], downcast="integer"
-                )
+            elif column_metadata.scale > 0 and not str(pandas_dtype).startswith(
+                "float"
+            ):
+                # For decimal columns, we want to cast it into float64 because pandas doesn't
+                # recognize decimal type.
+                pandas.to_numeric(pd_df[pandas_col_name], downcast="float")
+                if pd_df[pandas_col_name].dtype == "O":
```

Comment:
For my education, when is a pandas column dtype "0"?

Reply:
It is `O` not `0` LOL. It is when the dtype is an object.